### ISSUE TITLE
✍️ Contribution period: Mercy Birungi

### ISSUE THREAD

**Issue by MercybirungiS** [ersilia-os/ersilia#838]
### Week 1 - Get to know the community

- [x] Join the communication channels
- [x] Open a GitHub issue (this one!)
- [x] Install the Ersilia Model Hub and test the simplest model
- [x] Write a motivation statement to work at Ersilia
- [x] Submit your first contribution to the Outreachy site

### Week 2 - Install and run an ML model

- [x] Select a model from the suggested list
- [x] Install the model in your system
- [x] Run predictions for the EML
- [x] Compare results with the Ersilia Model Hub implementation!
- [x] Install and run Docker!

### Week 3 - Propose new models

- [x] Suggest a new model and document it (1)
- [x] Suggest a new model and document it (2)
- [x] Suggest a new model and document it (3)

### Week 4 - Prepare your final application

- [x] Submit the final application in the Outreachy website

**Comment by DhanshreeA** [comment]
Hi @MercybirungiS can you confirm if you could install Ersilia and run the simplest model and what is the output that you get?

**Comment by MercybirungiS** [comment]
Hi @DhanshreeA These are the steps I took to achieve step 3 in week 1 , Kindly review my output and let me know if I am on the right track 

**Task 3: Install the Ersilia Model Hub**

1. I set up the gcc compiler with this command 
`sudo apt install build-essential`

2.I installed miniconda3 with the following commands successfully 
```
mkdir -p ~/miniconda3
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O ~/miniconda3/miniconda.sh
bash ~/miniconda3/miniconda.sh -b -u -p ~/miniconda3
rm -rf ~/miniconda3/miniconda.sh
~/miniconda3/bin/conda init bash
~/miniconda3/bin/conda init zsh
```

3. I had Git and GitHubCLI configured
4. I set up Git LFS with these commands 
```
conda install git-lfs -c conda-forge
git-lfs install
```
5. I installed the Isaura data lake
```


conda activate ersilia
python -m pip install isaura==0.1
```
![ersiliaDownload](https://github.com/ersilia-os/ersilia/assets/79209607/c205ba76-d197-4c8a-9f85-98819db84ca8)

6.I had Docker set up on my device 
![dockerErsilia](https://github.com/ersilia-os/ersilia/assets/79209607/90a0bf2e-2843-49bd-9e28-b3fa779c7015)


7. I ran the following commands to install Ersilia
`conda create -n ersilia python=3.7`

8. I then ran the following commands to clone the Ersilia Python package 
`
![cloningersilia](https://github.com/ersilia-os/ersilia/assets/79209607/ef97795f-acff-46ec-9725-b06e7e55b3ad)



9.I then checked if Ersilia was working
![Uploading checkedErsiliaWorking.png…]()

10. Once I was sure ersilia was recognised in the CLI,  I then tested this model 

- First I fetched it with this command
`ersilia -v fetch eos2mrz`
![fetchedModelSuccessfully](https://github.com/ersilia-os/ersilia/assets/79209607/c6c3b834-abf1-4823-8bff-5721202b6f78)

- Then I ran `ersilia serve eos2mrz`
![image](https://github.com/ersilia-os/ersilia/assets/79209607/4dbf0699-fe78-44e9-99b7-ba1155487012)

![image](https://github.com/ersilia-os/ersilia/assets/79209607/0d78ebf6-14bf-44b9-865f-ab2c78444040)

- Then I ran this command and this was my out put . I ran it twice receiving the same output 

`ersilia -v api run -i "CCCC"`
![image](https://github.com/ersilia-os/ersilia/assets/79209607/c2e50fc0-ac1a-4634-b787-82a7e229daac)

**Comment by MercybirungiS** [comment]
**Motivation statement to work at Ersilia**

I am a Ugandan software engineer and data scientist passionate about leveraging technology to address real-world challenges. Ersilia's mission to provide data science tools to universities, hospitals, and laboratories in low-resourced countries resonates deeply with me.

My journey began as a software developer, where I honed my skills in creating user-centric applications. Over time, I discovered the potential of machine learning to extract valuable insights from data. Combining this with solid software engineering principles has been my focus.

I aim to bridge the gap between data science and software development, crafting innovative solutions that merge functionality with data-driven intelligence. My commitment to ongoing learning keeps me at the forefront of this evolving field.

Ersilia's work in making data science accessible to low-resource settings aligns with my vision. Machine learning can transform healthcare in these regions, but tools are often expensive and complex. Ersilia's open-source approach addresses this challenge.

I am enthusiastic about contributing to Ersilia's mission, bringing data science within reach for researchers in low-resource settings. With my background in software engineering and data science, I am well-suited for the Outreachy internship program. I am dedicated and eager to learn.

**Comment by MercybirungiS** [comment]
I am kindly requesting for guidance on how to submit my  first contribution to the Outreachy site if all is looking good . Thanks in Advance @DhanshreeA

**Comment by DhanshreeA** [comment]
Hi @MercybirungiS thank you for the updates. Please go ahead and record your first contribution on the Outreachy website.

**Comment by MercybirungiS** [comment]
**### Week 2 - Install and run an ML model**

**Comment by MercybirungiS** [comment]
Install the model in your system

I will be following the instructions on this github repository to install the model[]( https://github.com/Kohulan/Smiles-TO-iUpac-Translator)

**Comment by MercybirungiS** [comment]
Task 3 - Run predictions for the EML
I used this script to run the preditions 
`
from STOUT import translate_forward, translate_reverse

# SMILES to IUPAC name translation

SMILES = "CN1C=NC2=C1C(=O)N(C(=O)N2C)C"
IUPAC_name = translate_forward(SMILES)
print("IUPAC name of "+SMILES+" is: "+IUPAC_name)

# IUPAC name to SMILES translation

IUPAC_name = "1,3,7-trimethylpurine-2,6-dione"
SMILES = translate_reverse(IUPAC_name)
print("SMILES of "+IUPAC_name+" is: "+SMILES)
`
this script demonstrates how to use the STOUT (SMILES to IUPAC) model to perform two-way translations between SMILES notation and IUPAC chemical names. It translates a specific SMILES code into its corresponding IUPAC name and an IUPAC name into its corresponding SMILES code, and then displays the results in the command line.

This is the log I got for running the prediction 
[trainedStoutmodellog.txt](https://github.com/ersilia-os/ersilia/files/12907871/trainedStoutmodellog.txt)

The ouput basically showed that I performed two translations 

Running Translations:

 
 From a specific SMILES code, CN1C=NC2=C1C(=O)N(C(=O)N2C)C, to its IUPAC name, which resulted in "1,3,7-trimethylpurine-2,6-dione".

From a specific IUPAC name, "1,3,7-trimethylpurine-2,6-dione", back to its SMILES code, which resulted in CN1C=NC2=C1C(=O)N(C)C(=O)N2C.

The trained model was also downloaded on my laptop in a zipfile and folder , however cannot attach it here because if its big size

**Comment by MercybirungiS** [comment]
**Task 3 Run predictions for the EML**

**Comment by MercybirungiS** [comment]
Task 2 

I run predictions for  [Essential Medicines List](https://raw.githubusercontent.com/ersilia-os/ersilia/master/notebooks/eml_canonical.csv)

```
# This code  runs in Google Colab
import requests
from STOUT import translate_forward

def read_eml_csv(file_url):
   response = requests.get(file_url)
   response.raise_for_status()
   csv_data = response.text.splitlines()
   return list(csv.reader(csv_data))

def get_canonical_smiles_list(eml_list):
   return [name[2] for name in eml_list[1:]]

def translate_smiles_to_iupac(can_smiles_list):
   iupac_list = []
   for smiles in can_smiles_list:
       iupac = translate_forward(smiles)
       iupac_list.append(iupac)
   return iupac_list

# URL to the EML data CSV
eml_url = "https://raw.githubusercontent.com/ersilia-os/ersilia/master/notebooks/eml_canonical.csv"

# Clean the EML URL to remove any non-printable characters
eml_url = eml_url.encode("ascii", "ignore").decode()

# Read the EML data from the CSV URL
eml_list = read_eml_csv(eml_url)

# Extract the canonical SMILES strings from the EML data
can_smiles_list = get_canonical_smiles_list(eml_list)

# Translate the SMILES strings to IUPAC names
iupac_list = translate_smiles_to_iupac(can_smiles_list)

# Create a string with the predicted IUPAC names
results_str = "\n".join(iupac_list)

# Print the results to the console in Google Colab
display.Markdown(results_str)
```

**Comment by MercybirungiS** [comment]
The next task for me was to install the  [ADME@NCATS ](https://github.com/ncats/ncats-adme) model 
I followed the instructions on the READ.me page and this was my output

**Comment by MercybirungiS** [comment]
Task 3 Run Predictions 

I ran about four predictions , the screenshots and csv file outputs have been attached , with the output explanation 

First predition 
![image](https://github.com/ersilia-os/ersilia/assets/79209607/1d3e415d-6c23-4967-a92c-50d3dff66ca7)
The above screenshot shows  the predictions of a machine learning model for the liver microsomal and cytosolic stability of a molecule. The model predicts that the molecule is stable in the liver microsomal and cytosolic environments, with probabilities of 0.95 and 0.71, respectively.

Liver microsomal stability is a measure of how resistant a molecule is to metabolism by the liver microsomes. The liver microsomes are a group of enzymes that are responsible for metabolizing many drugs and other xenobiotics (foreign substances). If a molecule is unstable in the liver microsomes, it is likely to be metabolized quickly and excreted from the body.

Liver cytosolic stability is a measure of how resistant a molecule is to metabolism by the liver cytosol. The liver cytosol is the fluid that fills the liver cells. If a molecule is unstable in the liver cytosol, it is likely to be metabolized quickly and excreted from the body.

The predictions from the model suggest that the molecule is likely to be well-tolerated by the liver and have a long half-life. This is because the molecule is predicted to be stable in both the liver microsomal and cytosolic environments.

_csv file output_
[firstPrediction.csv](https://github.com/ersilia-os/ersilia/files/12934891/firstPrediction.csv)

Second predition
![image](https://github.com/ersilia-os/ersilia/assets/79209607/68ce4c05-05c3-4bb1-b639-45fabd777527)
_Interpretation_ 
This prediction is for the PAMPA permeability of a molecule at pH 5.0. The model predicts that the molecule has a low permeability, with a probability of 0.99.

This suggests that the molecule may not be well-absorbed from the gastrointestinal tract. This could be a problem for drug molecules, as it would limit their bioavailability (the amount of drug that reaches the systemic circulation).
_csv output_
[secondPrediction.csv](https://github.com/ersilia-os/ersilia/files/12935343/secondPrediction.csv)
 


Third prediction

![image](https://github.com/ersilia-os/ersilia/assets/79209607/0f29354f-91f1-4e58-9148-ef22667f5887)
_Interpretation_ 
This prediction is for the PAMPA permeability of a molecule at pH 7.4. The model predicts that the molecule has a low or moderate permeability, with a probability of 0.9.

This suggests that the molecule may not be well-absorbed from the gastrointestinal tract. However, the probability is slightly lower than for the prediction at pH 5.0, suggesting that the molecule may be slightly more permeable at pH 7.4.

_csv output_
[thirdPrediction.csv](https://github.com/ersilia-os/ersilia/files/12935414/thirdPrediction.csv)



Fourth prediction 
![image](https://github.com/ersilia-os/ersilia/assets/79209607/8bcfb5cd-fe49-44c6-af72-6e30d4ad3983)
_Interpretation_ 
This prediction is for the CYP450 enzyme inhibition potential of a molecule. The model predicts that the molecule is a weak inhibitor of the CYP3A4 enzyme, with a probability of 0.74.

This suggests that the molecule is unlikely to have a significant effect on the metabolism of other drugs that are metabolized by CYP3A4. However, it is important to note that this is just a prediction and the actual CYP3A4 inhibition potential of the molecule will need to be experimentally determined.

_csv output_
[fourthPrediction.csv](https://github.com/ersilia-os/ersilia/files/12935406/fourthPrediction.csv)

**Comment by MercybirungiS** [comment]
Task four : Compare results with the Ersilia Model Hub implementation!

For this part , i first checked out [Ersilia model hub](https://www.ersilia.io/model-hub)

**Comment by MercybirungiS** [comment]
Next I ran a prediction of the model on Ersilia with this command using the the _eml_canonical.csv_ as my input 
`ersilia -v api predict -i /home/mercy/eml_canonical.csv -o model_output.csv -l log.txt`

This is the output i got 
[ersiliamodeloutput.csv](https://github.com/ersilia-os/ersilia/files/12937526/ersiliamodeloutput.csv)

**Comment by MercybirungiS** [comment]
### Comparison of Predictions

### Ersilia Model Hub vs. First Prediction
I used excel sheets to compare the csv file for my First prediction and that for Ersilia Model hub and this is the output I came up with 
| Molecule | Ersilia Model Hub Prediction | First Prediction | Difference |
|---|---|---|---|
| SMILES: O=C(O)C1=C(C(=C(C1)C)C)CCl | Stable (0.95) | Stable (0.99) | 0.04 |
| SMILES: CCC(=O)NCCC(CCl)C | Stable (0.92) | Stable (0.97) | 0.05 |
| SMILES: CC(C)OCC(=O)N | Stable (0.87) | Stable (0.95) | 0.08 |
| SMILES: CC(=O)OC(C)CCl | Stable (0.85) | Stable (0.92) | 0.07 |
| SMILES: CCC(=O)OCC | Stable (0.82) | Stable (0.90) | 0.08 |

**Comment by MercybirungiS** [comment]
I had  docker  and docker compose installed on my laptop

I ran this command to confirm 
`docker version`
```
Client: Docker Engine - Community
 Cloud integration: v1.0.29
 Version:           24.0.6
 API version:       1.41 (downgraded from 1.43)
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:31:44 2023
 OS/Arch:           linux/amd64
 Context:           desktop-linux

Server: Docker Desktop 4.13.0 (89412)
 Engine:
  Version:          20.10.20
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       03df974
  Built:            Tue Oct 18 18:18:35 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.8
  GitCommit:        9cd3357b7fd7218e4aec3eae239db1f68a5a6ec6
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

```

**Comment by MercybirungiS** [comment]
I will then move on to week 3 , however @DhanshreeA kindly have a look at my week 2 tasks . Thankyou

**Comment by DhanshreeA** [comment]
And yes you can proceed with week 3 tasks!

**Comment by MercybirungiS** [comment]
> Hi @MercybirungiS many thanks for your contributions. Could you provide a csv with for the entire EML file with the same comparison as you have provided above? It will be very helpful for us, thank you!

Hi @DhanshreeA this is the entire excel doc , divided into different different sheets , first for the first pediction i made as seen above , the next one is for the Ersilia model , then the third is the comparison . I have also attached the comparison csv 
[Entire EML CSV.xlsx](https://github.com/ersilia-os/ersilia/files/13050586/Entire.EML.CSV.xlsx)
[Entire EML CSV - Comparison of the two outputs.csv](https://github.com/ersilia-os/ersilia/files/13050605/Entire.EML.CSV.-.Comparison.of.the.two.outputs.csv)

**Comment by MercybirungiS** [comment]
### Week 3 - Propose new models

In this week we are required to document 3 models 

these were the main instructions from the Ersilia book 

_"A big part of what we do at Ersilia is to screen the scientific literature in search of new models and datasets of interest to our community. We are always looking for models that can help speed up drug discovery against infectious and neglected diseases.
Find one publication that describes an open source ML model that could be of interest to Ersilia (activity against a specific pathogen, cytotoxicity, side effects...) and link it in the thread.
"_

**Comment by MercybirungiS** [comment]
### Model 1

# DeepChem

**Summary:**

DeepChem is a Python library that provides a variety of tools for machine learning in chemistry. It can be used to predict drug-target interactions, drug toxicity, and other properties of molecules.

**Why it would be relevant to Ersilia:**

DeepChem could be used by Ersilia to screen for new drug candidates that are active against specific pathogens. For example, Ersilia could train a DeepChem model to predict the binding affinity of different molecules to a protein target that is known to be involved in the pathogen's life cycle. Once the model is trained, Ersilia could use it to screen a large database of molecules for potential drug candidates.

**How to implement it:**

The DeepChem library is well-documented and easy to use. The authors of the paper also provide a tutorial on how to use DeepChem to predict drug-target interactions. To implement DeepChem, Ersilia would need to:

1. Install the DeepChem library
2. Download the dataset of drug-target interactions that they want to train the model on
3. Train the DeepChem model
4. Evaluate the model's performance on a held-out test set
5. Use the model to screen for new drug candidates

**Is the code ready to be used?**

Yes, the DeepChem library is ready to be used. The authors of the paper provide a variety of code examples, including an example of how to train a model to predict drug-target interactions.

**Is the underlying data available?**

Yes, the underlying data for the DeepChem datasets is available from the [DeepChem website.](https://deepchem.io/)

**Link to the code:**
[GitHub Repository](https://github.com/deepchem/deepchem)

**Link to the research paper:**
[Research Paper](https://pubs.acs.org/doi/10.1021/acs.jcim.9b00793)

**Comment by MercybirungiS** [comment]
### Model 2
# MoleculeNet

**Summary:**

MoleculeNet is a Python library for machine learning on molecules. It provides a variety of datasets and pre-trained models for tasks such as predicting drug-target interactions, drug toxicity, and ADME properties.

**Why it would be relevant to Ersilia:**

MoleculeNet could be used by Ersilia to train models for a variety of drug discovery tasks. For example, Ersilia could train a MoleculeNet model to predict the cytotoxicity of new drug candidates against a specific pathogen or to predict the drug-likeness of new drug candidates.

**How to implement it:**

The MoleculeNet library is well-documented and easy to use. The authors of the paper also provide a tutorial on how to use MoleculeNet to train models for drug discovery tasks. To implement MoleculeNet, Ersilia would need to:

1. Install the MoleculeNet library
2. Download the dataset of drug-target interactions or other properties that they want to train the model on
3. Train the MoleculeNet model
4. Evaluate the model's performance on a held-out test set
5. Use the model to make predictions for new drug candidates

**Is the code ready to be used?**

Yes, the MoleculeNet library is ready to be used. The authors of the paper provide a variety of code examples, including examples of how to train models for drug discovery tasks.

**Is the underlying data available?**

Yes, the underlying data for the MoleculeNet datasets is available from the MoleculeNet website.

**Link to the code:**
https://github.com/deepchem/moleculenet

**Link to the research paper:**
https://www.researchgate.net/publication/314182452_MoleculeNet_A_Benchmark_for_Molecular_Machine_Learning

**Comment by GemmaTuron** [comment]
Hello,

Thanks for your work during the Outreachy contribution period, we hope you enjoyed it! We will now close this issue while we work on the selection of interns. Thanks again!
